snuba/clusters/cluster.py

```diff
@@ -374,16 +375,49 @@ def __get_cluster_nodes(self, cluster_name: str) -> Sequence[ClickhouseNode]:
     if (s not in DEV_STORAGE_SETS or settings.ENABLE_DEV_FEATURES)
 }
 
-# Map all storages to clusters via storage sets
-_STORAGE_SET_CLUSTER_MAP = {
-    storage_set: cluster
-    for cluster in CLUSTERS
-    for storage_set in cluster.get_storage_set_keys()
-}
+_STORAGE_SET_CLUSTER_MAP: Optional[Dict[StorageSetKey, ClickhouseCluster]] = None
+
+
+def _get_storage_set_cluster_map() -> Dict[StorageSetKey, ClickhouseCluster]:
+    global _STORAGE_SET_CLUSTER_MAP
+    # Map all storages to clusters via storage sets
+    if _STORAGE_SET_CLUSTER_MAP is None:
+        _STORAGE_SET_CLUSTER_MAP = {
+            storage_set: cluster
+            for cluster in CLUSTERS
+            for storage_set in cluster.get_storage_set_keys()
+        }
+    return _STORAGE_SET_CLUSTER_MAP
+
+
+class UndefinedClickhouseCluster(ClickhouseCluster):
+    pass
 
 
 def get_cluster(storage_set_key: StorageSetKey) -> ClickhouseCluster:
+    """Return a clickhouse cluster for a storage set key.
+
+    If the storage set key is not defined
+    in the CLUSTERS config, it will return an UndefinedClickhouseCluster.
+
+    This allows not yet fully configured datasets to be deployed without causing
+    the whole server to crash. Only queries to that specific storage set will
+    fail (since its cluster is undefined)
+    """
     assert (
         storage_set_key not in DEV_STORAGE_SETS or settings.ENABLE_DEV_FEATURES
     ), f"Storage set {storage_set_key} is disabled"
-    return _STORAGE_SET_CLUSTER_MAP[storage_set_key]
+    return _get_storage_set_cluster_map().get(
+        storage_set_key,
+        UndefinedClickhouseCluster(
+            host="undefined_clickhouse_cluster",
+            port=420,
+            user="filippo",
+            password="p455w0rd",
+            database="nonya",
+            http_port=69,
+            storage_sets={storage_set_key.value},
+            single_node=False,
```

Comment:
??

Reply:
Does the docstring not explain it?

Reply:
Why can't we just return None, instead of this weird object?

Reply:
My initial quick take was about the hardcoded values, which I would strongly discourage, as somebody getting that result would try to make sense of. 

In design term, agree with Lyn. If you want to avoid the exception, returning an invalid value seems not to solve the problem and does not seem to be easier to manage than None. At least with None, the callsite knows they have to do something special for that case. The type checker will tell them.